be/src/vec/exec/format/parquet/arrow_memory_pool.h

```diff
@@ -0,0 +1,204 @@
+        // updating knowing that it already reached a value even higher than ours.
+        const auto allocated = old_bytes_allocated + size;
+        while (max_memory < allocated &&
+               !max_memory_.compare_exchange_weak(
+                       /*expected=*/max_memory, /*desired=*/allocated, std::memory_order_acq_rel)) {
+        }
+    }
+
+    inline void did_reallocate_bytes(int64_t old_size, int64_t new_size) {
+        if (new_size > old_size) {
+            did_allocate_bytes(new_size - old_size);
+        } else {
+            did_free_bytes(old_size - new_size);
+        }
+    }
+
+    inline void did_free_bytes(int64_t size) {
+        bytes_allocated_.fetch_sub(size, std::memory_order_acq_rel);
+    }
+};
+
+template <typename Allocator = ArrowAllocator>
+class ArrowMemoryPool : public arrow::MemoryPool {
+public:
+    ~ArrowMemoryPool() override = default;
+
+    arrow::Status Allocate(int64_t size, int64_t alignment, uint8_t** out) override {
+        if (size < 0) {
+            return arrow::Status::Invalid("negative malloc size");
+        }
+        if (static_cast<uint64_t>(size) >= std::numeric_limits<size_t>::max()) {
+            return arrow::Status::OutOfMemory("malloc size overflows size_t");
+        }
+        RETURN_NOT_OK(_allocator.allocate_aligned(size, alignment, out));
+#ifndef NDEBUG
+        // Poison data
+        if (size > 0) {
+            DCHECK_NE(*out, nullptr);
+            (*out)[0] = kAllocPoison;
+            (*out)[size - 1] = kAllocPoison;
+        }
+#endif
+
+        _stats.did_allocate_bytes(size);
+        return arrow::Status::OK();
+    }
+
+    arrow::Status Reallocate(int64_t old_size, int64_t new_size, int64_t alignment,
+                             uint8_t** ptr) override {
+        if (new_size < 0) {
+            return arrow::Status::Invalid("negative realloc size");
+        }
+        if (static_cast<uint64_t>(new_size) >= std::numeric_limits<size_t>::max()) {
+            return arrow::Status::OutOfMemory("realloc overflows size_t");
+        }
+        RETURN_NOT_OK(_allocator.reallocate_aligned(old_size, new_size, alignment, ptr));
+#ifndef NDEBUG
+        // Poison data
+        if (new_size > old_size) {
+            DCHECK_NE(*ptr, nullptr);
+            (*ptr)[old_size] = kReallocPoison;
+            (*ptr)[new_size - 1] = kReallocPoison;
+        }
+#endif
+
+        _stats.did_reallocate_bytes(old_size, new_size);
+        return arrow::Status::OK();
+    }
+
+    void Free(uint8_t* buffer, int64_t size, int64_t alignment) override {
+#ifndef NDEBUG
+        // Poison data
+        if (size > 0) {
+            DCHECK_NE(buffer, nullptr);
+            buffer[0] = kDeallocPoison;
+            buffer[size - 1] = kDeallocPoison;
+        }
+#endif
+        _allocator.deallocate_aligned(buffer, size, alignment);
+
+        _stats.did_free_bytes(size);
+    }
+
+    void ReleaseUnused() override { _allocator.release_unused(); }
```

Comment:
`je_mallctl(fmt::format("arena.{}.purge", MALLCTL_ARENAS_ALL)` is a very heavy and time-consuming operation.

if jemalloc dirty pages exceeding GB, this method may take more than 1 second to clear.
so be careful not to block the program.

Reply:
> `je_mallctl(fmt::format("arena.{}.purge", MALLCTL_ARENAS_ALL)` is a very heavy and time-consuming operation.
> 
> if jemalloc dirty pages exceeding GB, this method may take more than 1 second to clear. so be careful not to block the program.

Default arrow memory pool also defined this function.